kernel: Modify the signature of `k_mem_slab_free` via stable API change

carlescufi opened the pull request:
kernel: Modify the signature of `k_mem_slab_free()` via the [Stable API Change procedure](https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html#introducing-incompatible-changes)

Modify the signature of the `k_mem_slab_free()` function with a new one,
replacing the old `void **mem` with `void *mem` as a parameter.

Fixes #61888.

Stable API change RFC: #61937

Alternative to https://github.com/zephyrproject-rtos/zephyr/pull/61897

carlescufi committed:
kernel: Modify the signature of k_mem_slab_free()

Modify the signature of the k_mem_slab_free() function with a new one,
replacing the old void **mem with void *mem as a parameter.

The following function:
void k_mem_slab_free(struct k_mem_slab *slab, void **mem);

has the wrong signature. mem is only used as a regular pointer, so there
is no need to use a double-pointer. The correct signature should be:
void k_mem_slab_free(struct k_mem_slab *slab, void *mem);

The issue with the current signature, although functional, is that it is
extremely confusing. I myself, a veteran Zephyr developer, was confused
by this parameter when looking at it recently.

All in-tree uses of the function have been adapted.

Fixes #61888.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

gmarull commented:
nit: to preserve bisectability I guess all changes should be part of a single commit

carlescufi commented:
> nit: to preserve bisectability I guess all changes should be part of a single commit

I thought about it, I have no objections, let's hear others chime in.

andyross reviewed (dismissed):
This works for me too

henrikbrixandersen reviewed (dismissed):
Looks good. This is a tree-wide change and should go the appropriate process.

carlescufi commented:
> Looks good. This is a tree-wide change and should go the appropriate process.

Since this is also a stable API change I have instead followed [this process](https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html#introducing-incompatible-changes)

jukkar commented:
The related issue gives details why this change is done, but the motivation i.e., the question "why", is missing in the commit msg. It would be great if such details would be found also in the commit that introduces the change.

gmarull reviewed (dismissed):
remember to squash when merging

carlescufi commented:
> The related issue gives details why this change is done, but the motivation i.e., the question "why", is missing in the commit msg. It would be great if such details would be found also in the commit that introduces the change.

Good point. Done

de-nordic reviewed (dismissed):
MCUmgr looks OK

nashif commented on doc/releases/release-notes-3.5.rst:
```diff
@@ -20,6 +20,9 @@ API Changes
 Changes in this release
 =======================
 
+* The kernel :c:func:`k_mem_slab_free` function has changed its signature, now
```
My main concern is that this one liner in the release notes is not enough for someone using Zephyr to know the nature and impact of the change and the fact that they HAVE to change their code or risk their application changing behavior or crashing. IMO we need to isolate breaking API changes into a seperate document with more details and information about how migration to new API can be done (assisted by scripts, etc.)

carlescufi replied:
Instead of a separate document, how about a new section in the existing document, at the top, that says "Things you need to change in your code to migrate to this version"?

carlescufi replied:
Actually, I think you are right. Let's just introduce a new migration document for each release. I will send a Pull Request for this.

carlescufi committed:
kernel: mem_slab: Add assert to k_mem_slab_free()

Add an assert to ensure the pointer provided by the user points to one
of the available blocks in the slab.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi commented:
> Was there any discussion of leaving the `k_mem_slab_free()` as is and introducing another `k_mem_slab_freeXXX()` (replace the XXX with your favorite postfix) that would exist together with the old variant. The other could call the other with suitable pointer magic. This way we would not need to worry about out of tree use of the older function as it would be found in the sources. The documentation could encourage use of the newer variant of the function and there would be no need to deprecate the old one so that the out of tree use of that could continue.

Yes, please see this: https://github.com/zephyrproject-rtos/zephyr/pull/61897

jukkar commented:
> Yes, please see this: #61897

Sure, I was just thinking to **not** deprecate the old API but leave it there, and introduce 2nd variant.

nordicjm commented:
> > Yes, please see this: #61897
> 
> Sure, I was just thinking to **not** deprecate the old API but leave it there, and introduce 2nd variant.

The issue isn't the deprecation, the issue is the second function which no-one wants

carlescufi commented:
In this last push, the following changes have been made:

- Squashed all commits together for bisectability
- Added a new `_ASSERT()` statement to check the pointer being passed in
- Moved the info about the change to the new migration guide

de-nordic reviewed (approved):
MCUmgr is ok.

andyross reviewed (approved):
refresh +1
